### PR TITLE
Submission freezing

### PR DESCRIPTION
**Pull request by sujnesh**

### CODE REVIEW

File: app/controllers/submissions_controller.rb [hunk 1]
```diff
@@ -61,7 +61,8 @@ def index
     @search = filter.search(search_params)
     @search.sorts = 'created_at desc' if @search.sorts.empty?
     params[:page] ||= 1
-    @submissions  = @search.result.includes(:participant).per_page_kaminari(params[:page]).per(10)
+    filter = @challenge.submission_filter
```
Comment (skbly7) [inline app/controllers/submissions_controller.rb:64]: why is this happening in `index`? it should be only for `lock`?

Comment (sujnesh) [reply app/controllers/submissions_controller.rb:64]: Oh, I misunderstood it. I'll correct it.

File: app/controllers/submissions_controller.rb [hunk 2]
```diff
@@ -190,6 +191,58 @@ def new_api
...
+          LockedSubmission.create!(
+            submission_id: params[:locked_submission][:submission_id].to_i,
+            challenge_id: @challenge.id,
+            locked_by: current_participant.id
+            )
+        end
+      end
+    end
+
+    return redirect_to(root_path, flash: { success: 'We have received your submission.' })
```
Comment (sujnesh) [inline app/controllers/submissions_controller.rb:222]: Redirect to same page

File: app/controllers/submissions_controller.rb [hunk 3]
```diff
@@ -420,4 +473,25 @@ def render_or_json(render)
       render render
     end
   end
+
+  def create_default_locked_submissions
+    participant_ids = LockedSubmission.where(challenge_id: @challenge.id).pluck(:locked_by)
+    challenge_participant_ids = @challenge.submissions.pluck(:participant_id).uniq
```
Comment (sujnesh) [inline app/controllers/submissions_controller.rb:479]: add contition grading status cd = 'graded'

File: app/controllers/submissions_controller.rb [hunk 4]
```diff
@@ -420,4 +473,25 @@ def render_or_json(render)
...
+    remaining_participants = challenge_participant_ids - participant_ids
+
+    Participant.where(id: remaining_participants).each do |participant|
+      next if LockedSubmission.where(locked_by: participant.id).exists?
+
+      team = participant.teams.where(challenge_id: @challenge.id).first
+      if team.blank?
+        submission = @challenge.submissions.where(participant_id: participant.id).order(@challenge.submission_freezing_order).first
+        LockedSubmission.create!(submission_id: submission.id, locked_by: participant.id, challenge_id: @challenge.id) if submission
```
Comment (sujnesh) [inline app/controllers/submissions_controller.rb:488]: Don't create locked_submission for default

**Pull request merged**